Automattic/wp-calypso
# Jetpack Backup: Correctly display backup status and navigation for sites with lots of activities

Fixes `1164141197617539-as-1186890594349665`, `1184358343400545-as-1190849472047377`.

#### Changes proposed in this Pull Request

* Split the Backup page into several discrete components, like `DatePicker`, `BackupStatus`, and `SearchResults`.
* Create separate status components for daily and real-time backups, because the information they display is fairly diverged.
* Preserve feature flag behavior for `jetpack/backup-simplified-screens-i4`, but move it under `backup/status/simplified-screens-i4.jsx`.
* Create new custom hooks `useDailyBackupStatus` and `useRealtimeBackupStatus`, to simplify the process of getting the information to display the backup status for a particular site on a given date.
* Try to make all the logic -- and there's so much of it -- more self-contained and easy to reason about.
* Move a couple styles to get rid of linting errors as a result of all the above changes.
* Make `requestActivityLogs` respect and use the `number` param to control how many results come back.

#### Implementation notes

* You may notice some brief flickering as activity log information loads. I'm not a fan of this, and I tried to mitigate it by eagerly fetching data for 1 day before/after the selected date. This solution works, mostly, but I feel like it's a bit quick-n-dirty.

#### Testing instructions

**NOTE:** This pull request touches pretty much every part of the Backup page and its `simplified-screens-i4` variant! Be extra thorough; keep an eagle eye out for issues with styling and behavior; and importantly, remember to test both Calypso and Jetpack Cloud on both mobile and desktop layouts.

Most of these instructions are for targeted testing against known possible regressions, but we have one scenario we're trying to fix with this PR: **sites that have many Activity Log events** should be able to see their backups and navigate forward/backward through time without difficulty. We should be showing details of their changes on the page, just like sites with fewer activities (though we're limiting to only the 1000 latest per day at the moment).

If you need to gain access to a site with lots of activities in its Activity Log, message me privately and I can set you up. I recommend testing this PR side-by-side with production, because other than the bug-fix case described in the last paragraph, everything should behave the same as it does today.

For sites with either Backup Daily or Real-time:
* If today's backup failed but a successful one exists in the past, a link is displayed to that successful backup (see #46886).
* The date picker should be unrestricted, allowing navigation to any calendar date, **especially** for sites with lots of activities.
* All dates are correct for the site's time zone, and all "deltas" display appropriately with respect to the state of the `jetpack/backup-simplified-screens-i4` feature flag.

For sites with Backup Daily, verify that:
* The daily backup status correctly displays the status of the daily backup attempt that happened (or didn't happen) on the selected date.

For sites with Backup Real-time, verify that:
* The daily backup status correctly displays the most recent rewindable activity as the backup for the selected date. If no activities are available on the selected date, behavior should be similar to sites with Backup Daily re: success, failure, etc.

#### Screenshots

<img width="742" alt="Screen Shot 2020-10-29 at 14 55 59" src="https://user-images.githubusercontent.com/670067/97625727-e20bb780-19f6-11eb-9fc9-c884f0021524.png">
<img width="744" alt="Screen Shot 2020-10-29 at 15 08 47" src="https://user-images.githubusercontent.com/670067/97626981-b25daf00-19f8-11eb-9878-1958372c4f2e.png">
